apps/rush-lib/src/logic/DependencyIntegrator.ts

```diff
@@ -0,0 +1,355 @@
+import * as colors from 'colors';
+import * as path from 'path';
+import * as semver from 'semver';
+
+import {
+  JsonFile,
+  FileConstants,
+  IPackageJson
+} from '@microsoft/node-core-library';
+
+import { RushConfiguration } from '../api/RushConfiguration';
+import { InstallManager, IInstallManagerOptions } from './InstallManager';
+import { RushConfigurationProject } from '../api/RushConfigurationProject';
+import { VersionMismatchFinder } from '../api/VersionMismatchFinder';
+import { PurgeManager } from './PurgeManager';
+import { Utilities } from '../utilities/Utilities';
+
+/**
+ * The type of SemVer range specifier that is prepended to the version
+ */
+export const enum SemVerStyle {
+  Exact = 'exact',
+  Caret = 'caret',
+  Tilde = 'tilde'
+}
+
+/**
+ * The type of dependency that this is. Note: we don't support PeerDependencies
+ */
+export const enum DependencyKind {
+  DevDependency = 'devDependency',
+  Dependency = 'dependency'
+}
+
+/**
+ * Configuration options for adding or updating a dependency in a single project
+ */
+export interface IUpdateProjectOptions {
+  /**
+   * The project which will have its package.json updated
+   */
+  project: RushConfigurationProject;
+  /**
+   * The name of the dependency to be added or updated in the project
+   */
+  packageName: string;
+  /**
+   * The new SemVer specifier that should be added to the project's package.json
+   */
+  newVersion: string;
+  /**
+   * The type of dependency that should be updated. If left empty, this will be auto-detected.
+   * If it cannot be auto-detected an exception will be thrown.
+   */
+  dependencyKind?: DependencyKind;
+  /**
+   * If specified, the package.json will only be updated in memory, but the changes will not
+   * be written to disk.
+   */
+  doNotSave?: boolean;
```

Comment:
This is weird:

1. If RushConfiguration is mutable, then it needs a RushConfiguration.saveToDisk() and RushConfiguration.revertChanges() or RushConfiguration.reload() to manage the modified state where it has become out of sync with the files on disk.

2. If RushConfiguration is a readonly cache, then PackageJsonUpdater needs to maintain its own state, and we need some way to invalidate/reload RushConfiguration after the state has changed.